thirdparty/papertrail.go

```diff
@@ -0,0 +1,127 @@
+package thirdparty
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/pkg/errors"
+)
+
+type PapertrailSpan struct {
+	ID        string `json:"id"`
+	SHA256sum string `json:"sha256sum"`
+	Filename  string `json:"filename"`
+	Build     string `json:"build"`
+	Platform  string `json:"platform"`
+	Product   string `json:"product"`
+	Version   string `json:"version"`
+	Submitter string `json:"submitter"`
+	KeyID     string `json:"key_id"`
+	Time      uint64 `json:"time"`
+}
+
+type PapertrailProductVersion struct {
+	Product string           `json:"product"`
+	Version string           `json:"version"`
+	Spans   []PapertrailSpan `json:"spans"`
+}
+
+type PapertrailClient struct {
+	*http.Client
+	Address   string
+	KeyID     string
+	SecretKey string
+}
+
+type TraceArgs struct {
+	Build     string
+	Platform  string
+	Filename  string
+	Sha256    string
+	Product   string
+	Version   string
+	Submitter string
+}
+
+func (c *PapertrailClient) Trace(ctx context.Context, args TraceArgs) error {
+	params := url.Values{}
+
+	params.Set("build", args.Build)
+	params.Set("platform", args.Platform)
+	params.Set("filename", args.Filename)
+	params.Set("sha256", args.Sha256)
+	params.Set("product", args.Product)
+	params.Set("version", args.Version)
+	params.Set("submitter", args.Submitter)
+
+	addr := fmt.Sprintf("%s/trace?%s", c.Address, params.Encode())
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, addr, nil)
+	if err != nil {
+		return errors.Wrap(err, "new request")
+	}
+
+	req.Header.Set("X-PAPERTRAIL-KEY-ID", c.KeyID)
+	req.Header.Set("X-PAPERTRAIL-SECRET-KEY", c.SecretKey)
+
+	res, err := c.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "do post")
+	}
+
+	defer res.Body.Close()
+
+	b, err := io.ReadAll(res.Body)
+	if err != nil {
+		return errors.Wrap(err, "read body")
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return errors.Errorf("status %d: %s", res.StatusCode, string(b))
+	}
+
+	return nil
+}
+
+func (c *PapertrailClient) GetProductVersion(ctx context.Context, product, version string) (*PapertrailProductVersion, error) {
+	params := url.Values{}
+
+	params.Set("product", product)
+	params.Set("version", version)
+	params.Set("format", "json")
+
+	addr := fmt.Sprintf("%s/product-version?%s", c.Address, params.Encode())
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, addr, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "new request")
+	}
+
+	res, err := c.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "do get")
+	}
+
+	defer res.Body.Close()
+
+	b, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "read body")
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("status %d: %s", res.StatusCode, string(b))
```

Comment:
nit: we usually try to make error codes a bit more descriptive, in case the return isn't clear, like "getting papertrail product version". Also we have a utility helper function for this, util.RespErrorf
https://github.com/evergreen-ci/evergreen/blob/908d6545b88193c1c7a66886e332dd66c2a2bcb6/rest/client/methods.go#L818

Reply:
I went ahead and used this utility function, and changed the error messages